### PR TITLE
Add mlock when reading symbols in index reader

### PR DESCRIPTION
**Pull request by alexqyle**
Use mlock for Unix binary when reading symbols in index reader to lower page IO if symbols size is huge.

### CODE REVIEW

**Comment by alexqyle** [comment]
@codesome Please take a look at this PR. Feel free to leave any comments or concerns.

**Review by jesusvazquez — COMMENTED** [review]
Hello @alexqyle 

Could you add a benchmark to check the impact of these changes?

**Comment by roidelapluie** [comment]
We have looked at this pull request during our bug scrub.

@alexqyle friendly ping again, do you still plan to work on this?

Thank you for your contribution.

**Comment by roidelapluie** [comment]
Can you rebase this so we can run prombench ?

**Comment by beorn7** [comment]
My high level concerns here are the following:

This will effectively be a no-op if there is no memory pressure. (And I'm afraid that most benchmarks won't simulate memory pressure so we won't see an effect.)

The more important question is what happens if we _actually_ have memory pressure. Then this will keep the symbol table in RAM for sure, but is that the right call? If we are under memory pressure, many other mmap'd parts might also benefit from being locked in RAM. And if we lock too much in RAM, we will OOM, which is the very thing that the whole mmap approach tries to avoid. (At least that's my current understanding. Needs verification.)

In short, I think we need to discuss the rationale behind this change in more depth. And this is non-trivial to benchmark.

**Comment by bboreham** [comment]
See also https://github.com/prometheus/prometheus/pull/12587 which would cache label values in Go heap.  (Label names are already pre-emptively cached).

**Comment by krajorama** [comment]
Hello from the bug scrub: @alexqyle do you have any updates to the comments raised by others?

**Comment by bboreham** [comment]
Hello from the bug-scrub!
We accept @beorn7's argument that this is not desirable.
Thanks for your interest in the Prometheus project.
